server/src/main/scala/org/http4s/server/middleware/GZip.scala

```diff
@@ -24,13 +26,14 @@ object GZip {
             case resp: Response =>
               if (isZippable(resp)) {
                 logger.trace("GZip middleware encoding content")
-                // Need to add the Gzip header
+                // Need to add the Gzip header and trailer
+                val gzipTrailer = trailer(resp.body)
                 val b = chunk(header) ++
                   resp.body.through(deflate(
                     level = level,
                     nowrap = true,
                     bufferSize = bufferSize
-                  ))
+                  )) ++ gzipTrailer
```

Comment:
This concatenates `resp.body` to itself, effectively running the body twice.  That's fine if it's an in-memory stream, but if it's coming over the network, the second pass will be empty.

I would look into making a sink that calculates the crc as the bytes pass through.  Then you could do `header ++ resp.body.observe(crcSink) ++ trailer(crc)`.  It would also avoid folding the entire body into memory, as line 74 does.